flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/MapRangePartitioner.java

```diff
@@ -0,0 +1,373 @@
+    Map<Integer, Pair<Long, Integer>> assignmentInfo = Maps.newTreeMap();
+    assignment.forEach(
+        (key, keyAssignment) -> {
+          for (int i = 0; i < keyAssignment.assignedSubtasks.length; ++i) {
+            int subtaskId = keyAssignment.assignedSubtasks[i];
+            long subtaskWeight = keyAssignment.subtaskWeightsExcludingCloseCost[i];
+            Pair<Long, Integer> oldValue = assignmentInfo.getOrDefault(subtaskId, Pair.of(0L, 0));
+            assignmentInfo.put(
+                subtaskId, Pair.of(oldValue.first() + subtaskWeight, oldValue.second() + 1));
+          }
+        });
+
+    return assignmentInfo;
+  }
+
+  private Map<SortKey, KeyAssignment> buildAssignment(
+      int numPartitions,
+      NavigableMap<SortKey, Long> sortedStatistics,
+      long targetWeightPerSubtask,
+      long closeFileCostInWeight) {
+    Map<SortKey, KeyAssignment> assignmentMap =
+        Maps.newHashMapWithExpectedSize(sortedStatistics.size());
+    Iterator<SortKey> mapKeyIterator = sortedStatistics.keySet().iterator();
+    int subtaskId = 0;
+    SortKey currentKey = null;
+    long keyRemainingWeight = 0L;
+    long subtaskRemainingWeight = targetWeightPerSubtask;
+    List<Integer> assignedSubtasks = Lists.newArrayList();
+    List<Long> subtaskWeights = Lists.newArrayList();
+    while (mapKeyIterator.hasNext() || currentKey != null) {
+      // This should never happen because target weight is calculated using ceil function.
+      if (subtaskId >= numPartitions) {
+        LOG.error(
+            "Internal algorithm error: exhausted subtasks with unassigned keys left. number of partitions: {}, "
+                + "target weight per subtask: {}, close file cost in weight: {}, data statistics: {}",
+            numPartitions,
+            targetWeightPerSubtask,
+            closeFileCostInWeight,
+            sortedStatistics);
+        throw new IllegalStateException(
+            "Internal algorithm error: exhausted subtasks with unassigned keys left");
+      }
+
+      if (currentKey == null) {
+        currentKey = mapKeyIterator.next();
+        keyRemainingWeight = sortedStatistics.get(currentKey);
+      }
+
+      assignedSubtasks.add(subtaskId);
+      if (keyRemainingWeight < subtaskRemainingWeight) {
+        // assign the remaining weight of the key to the current subtask
+        subtaskWeights.add(keyRemainingWeight);
+        subtaskRemainingWeight -= keyRemainingWeight;
+        keyRemainingWeight = 0L;
+      } else {
+        // filled up the current subtask
+        long assignedWeight = subtaskRemainingWeight;
+        keyRemainingWeight -= subtaskRemainingWeight;
+
+        // If assigned weight is less than close file cost, pad it up with close file cost.
+        // This might cause the subtask assigned weight over the target weight.
+        // But it should be no more than one close file cost. Small skew is acceptable.
+        if (assignedWeight <= closeFileCostInWeight) {
```

Comment:
So, instead of not assigning key if there is not enough weight left for the current task, we push a bit more there to warrant opening a new file?

How do we handle if the current key doesn't have enough weight left for the next task?

Reply:
Oh... I see, we ignore those

Reply:
this might lead to some inaccuracy in weight calculation. E.g., assuming the key weight is 2 
 and close file cost is 2. key weight with close cost is 4. Let's assume the previous task
has a weight of 3 available. So weight of 3 for this key is assigned to the task and the 
residual weight of 1 is dropped. Then the routing weight for this key is 1 (minus the close
file cost), which is inaccurate as the accurate weight should be 2.

I thought about add the residual weight to the previous assignment. but it is also not alway accurate.
e.g. key weight is 11, target task weight before close cost is 10. so this key should be split into 2 files.
assuming close file cost is 1. key weight with close cost would be 13 (11 + 1x2). let's say the target 
task weight with close cost is 12. with add-back, the task would be assigned with weight of 13. 
routing weight would be 12 (13 - 1 close file cost). that is also inaccurate. With this simple greedy heuristic,
there is always some inaccuracy one way or the other.

but the inaccuracy should be small and doesn't skew the traffic distribution much.

Reply:
3 things to consider:
- If we merge, we might prefer merging files instead of splitting out (to help future readers)
- If the key weight is very small we might end up removing it altogether 
- We might prefer you simpler algorithm to handle all of the edge cases mentioned above

Will try to take another, more serious look at this soon